docs/poc2023q1/outline.md

```diff
@@ -0,0 +1,257 @@
+  ReplicaSet or Pod objects associated with a given Deployment
+  object).
+- A hierarchy with more than two levels.
+- More than baseline security (e.g., HTTPS, Secret objects, bearer
+  token based service authorization).
+
+It is TBD whether the implementation will support intermittent
+connectivity.  This depends on whether we can quickly and easily get a
+syncer that creates the appropriately independent objects in the edge
+cluster and itself tolerates intermittent connectivity.
+
+Also: initially this PoC will not transport non-namespaced objects.
+CustomResourceDefinitions are thus among the objects that will not be
+transported.  Transport of non-namespaced objects may be added after
+the other goals are achieved.
+
+As further matters of roadmapping development of this PoC:
+customization may be omitted at first, and summarization may start
+with only a limited subset of the implicit functionality.
+
+This PoC builds on TMC and makes some compromises to accommodate that.
+The implementation involves workload components (syncers) writing
+status information to inventory objects (SyncTargets).
+
+This PoC compromises on the connection initiative constraint in one
+place: the bootstrap step where central automation connects to edge
+clusters to install EMC machinery.
+
+## Design overview
+
+In very brief: the design is to reduce each edge placement problem to
+many instances of kcp's TMC problem.
+
+See [the overview picture](Edge-PoC-2023q1.svg) for an overview
+picture.
+
+## Inventory Management workspaces
+
+This design takes as a given that something maintains some kcp
+workspaces that contain dynamic collections of Location and SyncTarget
+objects as defined in [kcp
+TMC](https://github.com/kcp-dev/kcp/tree/main/pkg/apis), and that one
+view can be used to read those Location objects and one view can be
+used to read those SyncTarget objects.
+
+To enable the workload management layer to manipulate each edge
+cluster, that cluster's SyncTarget has an annotation that refers to a
+Secret that holds a kubeconfig that can be used by the Bootstrapper to
+install the edge-mc machinery in the edge cluster.  These annotations
+and Secrets are created by the inventory management layer for use by
+the workload management layer.
+
+To complete the plumbing of the syncers, for each SyncTarget the
+mailbox controller creates the following needed accompanying items in
+that SyncTarget's workspace (which must authorize the mailbox
+controller to do this).
+
+1. A ServiceAccount that the syncer will authenticate as.
+2. A ClusterRole manipulating that SyncTarget and the
+   APIResourceImports (what are these?).
+3. A ClusterRoleBinding that links that ServiceAccount with that
+   ClusterRole.
+
+FYI, those are the things that `kubectl kcp workload sync` directly
+creates besides the SyncTarget.
+
+## Edge Service Provider workspace
+
+The edge multi-cluster service is provided by one workspace that
```

Comment:
Is the "one" workspace a point in time statement or needed for this architecture?  I suspect that with large scale + sharding you may have multiple such workspaces?

Reply:
This workspace does store a data volume that goes up with usage, so I do not see a need to scale out this workspace.
It *does* have controllers whose workload scales in volume and throughput as usage increases, but that is a matter that can be handled by sharding those controllers, not this workspace.

Reply:
Doesn't a workspace reflects an API entry point and at the end an apiserver instance?  I was talking about throughput of api calls (but maybe I didn't get exactly what it is used for)

Reply:
Ah yes, this workspace has a view whose throughput (and data content volume) goes up with usage.

The scope of this PoC does not include sharding for scale.  In this PoC, there will be a single edge service provider workspace, and we will live with the scale implications of that.

Reply:
I tried to make this scope limitation clearer in the latest revision.